docs: adds docs for cutomizing search pages (#113)
inveniosoftware/invenio-search-ui: docs/customizing.rst

..
    This file is part of Invenio.
    Copyright (C) 2020 CERN.

    Invenio is free software; you can redistribute it and/or modify it
    under the terms of the MIT License; see LICENSE file for more details.

Customizing
===========

> how to inject/override the Invenio config (url, facets, the DOM id where to render the app): via parametrize or remove the Jinja macro and inject JSON config

Reviewer: minor: due to the structure of the sentence it is a bit hard to grasp:

```suggestion
There are two options to customize the configuration of the search pages (URL, facets, DOM ids, etc.). Parametrizing the [...] or removing the jinja macro and injecting the JSON config into the bare element [...]
```

Basically, I read that and I honestly did not understand what was "parameterize" or the "JSON config" (of what?) a few code examples might shed some light (most likely references to the content below is enough).

Author: Ah, well this was a leftover from the original issue. I was intending to actually skip it. Do you feel we should add this since it is also covered later in the docs?

Reviewer: I think quick introductions are always good, to have a fast view of whats in there. For example:

*In section one we cover this and in section two this this and this*

Mainly to give the user the power to choose were to go, being informed.

Author: In this case i feel like the introduction is not that needed since the segments are relatively short with a descriptive enough title, also I would avoid putting this much information at the start of the documentation since the user doesn't have the tools (architectural knowledge contained in the rest of the doc) to decompress them.

Configuring the settings of your Search Application
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

A helper is provided to generate the configuration needed for the
React-Searchkit application.
This configuration contains the following information:

> `aggs`: a list of aggregations that will add accordingly facets to your search.

Reviewer: ```suggestion
`aggs`: A list of aggregations that will accordingly add facets to your search.
```

Author: ```suggestion
`aggs`: a list of aggregations that will add facets to your search accordingly.
```
Maybe this would be even more appropriate.

.. code-block:: python

    [
        {
            "aggName": "access_right",
            "field": "access_right",
            "title": "Access Right"
        }
    ]

> `appId`: the value here is needed since it will be used internally to

Reviewer: minor: I wonder how this display. Is it intended to be a list? e.g.:

- aggs
- appId
- ...

Author: ![Screenshot 2020-12-04 at 17 27 09](https://user-images.githubusercontent.com/16304052/101188407-f8d8a600-3655-11eb-9a61-1a08946437a2.png)

> `appId`: the value here is needed since it will be used internally to

Reviewer: ```suggestion
`appId`: The value here is needed since it will be used internally to
```

differentiate the resources of multiple apps.
"rdm-search"
`initialQueryState`: This object is used to compose the initial search request
that will be sent to the backend.

.. code-block:: python

    {
        "hiddenParams": null,
        "size": 10
    }

> `layoutOptions`: here we can choose which layouts are going to be enabled

Reviewer: ```suggestion
`layoutOptions`: Here we can choose which layouts are going to be enabled
```

in our search page, if both are present then a layout switcher will also be

> present in the UI.

Reviewer: question: Is it automatic? IIRC you need to add the layourSwitcher component otherwise it just does not display. If so, it should be stated that the component *has to be added by the user*. (default templates have it though)

Reviewer: ping

Author: It's automatic, this is done [here](https://github.com/inveniosoftware/invenio-search-ui/blob/44b6122d8f066a92344f1c1a58b4d18e888b690c/invenio_search_ui/assets/semantic-ui/js/invenio_search_ui/components/Results.js#L30)

.. code-block:: python

    {
    "gridView": false,
    "listView": true
    }

`paginationOptions`: By changing this object you can increase the available
results per page.

.. code-block:: python

    {
    "defaultValue": 10,
    "resultsPerPage": [
      {
        "text": "10",
        "value": 10
      },
      {
        "text": "20",
        "value": 20
      }]
      }

> `searchApi`: the searchApi defines the URL for the list REST API, as well as

Reviewer: ```suggestion
`searchApi`: The searchApi defines the URL for the list REST API endpoint, as well as
```

> any headers that may need to be attached to the request.

Reviewer: ```suggestion
any headers that may need to be sent along with the request.
```

Reviewer: ping

.. code-block:: python

    {

>     "axios": {
>       "headers": {
>         "Accept": "application/vnd.inveniordm.v1+json"
>       },
>       "url": "/api/records",
>       "withCredentials": true
>     }

Reviewer: ```suggestion
        "axios": {
            "headers": {
                "Accept": "application/vnd.inveniordm.v1+json"
             },
            "url": "/api/records",
            "withCredentials": true
        }
```

Reviewer: merely aesthetic, feel free to resolve it.

    }

`sortOptions`

.. code-block:: python

    [
        {
        "sortBy": "bestmatch",
        "text": "Best match"
        },
        {
        "sortBy": "newest",
        "text": "Newest"
        }
    ]

This helper needs to be provided with the ID of the endpoint registered in your

> Invenio-Records-REST configuration.
> https://github.com/inveniosoftware/invenio-records-rest/blob/027794e6ae8dbe2448c59a64b30cdd3c4abb7eb7/invenio_records_rest/config.py#L26

Reviewer: ```suggestion
`Invenio-Records-REST configuration. <https://github.com/inveniosoftware/invenio-records-rest/blob/027794e6ae8dbe2448c59a64b30cdd3c4abb7eb7/invenio_records_rest/config.py#L26>`_
```

Author: So this was something that I wanted feedback on, is it valid to point to the code of another repo?
Since this will be a permalink to the current version.

Reviewer: I don't really like it but:
1- we need it
2- Better to be hyperlink than a full link (long an ugly) in the docs...

.. code-block:: python

    SearchAppInvenioRestConfigHelper.generate(
                {'endpoint_id': 'recid'})

It will then generate a react-searchkit compatible configuration, like the
one mentioned above.
You can provide this configuration to the jinja template via the
search_app_helpers macro.
You can also make adjustments to the default values
(e.g. page size) by providing your own when initialising the helper object.

.. code-block:: python

    SearchAppInvenioRestConfigHelper({'default_size':10}).generate(
                {'endpoint_id': 'recid'})

If a part of the automatically generated configuration doesn't suit your case

> you can completely overwite it by supplementing the .generate function with the

Reviewer: ```suggestion
you can completely overwrite it by supplying the .generate function with the
```

correct one.
This helper is provided to cover most of the basic cases, but if you feel
that you need to replace many parts of the generated configuration object

> you can also remove it and provide directly the configuration object.

Reviewer: ```suggestion
you can also remove it and provide the configuration object directly.
```

.. code-block:: python

    # For instance if you want to disable the facets in the UI despite
    # the aggregations being available in the backend.
    SearchAppInvenioRestConfigHelper({'default_size':10}).generate(
                {'endpoint_id': 'recid'}, {'aggs':[]})

Then just add it to your jinja template rendering the search page:

.. code-block:: html+jinja

    <div data-invenio-search-config='{{
    search_app_helpers.invenio_records_rest.generate(
        dict(
        endpoint_id="recid",
        app_id="search"
        )
    ) | tojson(indent=2) }}'></div>

> Configuring the components of your Search Application

Reviewer: comment: this section should go first (IMO), since it clearly explains the first most paragraph of the file, which otherwise is left a bit like 🤔 💭 (see first comment).

Author: I think we should remove this since it was a leftover.
I feel like this should come second since in terms of complexity it is more standard to need to change the configuration than override components.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To render a search page you need 3 elements, a jinja template containing a div
element with the attribute 'data-invenio-search-config' populated with a
react-searchkit configuration, while also importing an initialised SearchApp
from invenio-search-ui.

To do the latter you can create a `index.js` file under the `semamantic-ui/js`
folder and use the `createSearchAppInit` from invenio-search-ui to initialise a
SearchApp.
You can point to this additional file by including it to the `entry` object in
`webpack.py`.

.. code-block:: python

    entry={
        "invenio_search_ui_app": "./js/invenio_search_ui/app.js",
    },

Finally just include the created bundle in your jinja template:

.. code-block:: html+jinja

    {%- block javascript %}

>     {{ super() }}
>     {{ webpack['invenio_search_ui_app.js'] }}

Reviewer: major: This order has to be changed due to how `react-overridable` works. We had a big problem in the cookie-cutter, see [here](https://github.com/inveniosoftware/cookiecutter-invenio-instance/pull/253/files#diff-4842c5a124984dcd66aea9eaaa191570d428ca6e7898619498056573ae2f53c7R6). Basically, the overrides are loaded early (instead of lazy) so if super is called first, you *cannot override anything. See [here](https://github.com/inveniosoftware/invenio-search-ui/blob/master/invenio_search_ui/assets/semantic-ui/js/invenio_search_ui/components/SearchApp.js#L62).
```suggestion
    {{ webpack['invenio_search_ui_app.js'] }}
    {{ super() }}
```

..
    This file is part of Invenio.
    Copyright (C) 2020 CERN.

    Invenio is free software; you can redistribute it and/or modify it
    under the terms of the MIT License; see LICENSE file for more details.

Customizing
===========

> 

Reviewer: I think quick introductions are always good, to have a fast view of whats in there. For example:

In section one we cover this and in section two this this and this

Mainly to give the user the power to choose were to go, being informed.
